### DISCUSSION TITLE
Add "Max Lights per object" or "Per Cluster" option to Vulkan backend for mobile platforms

### DISCUSSION THREAD

**Discussion by luiscesjr** [godotengine/godot-proposals#7126] [3D]
### Describe the project you are working on

I'm working on a project game for mobile and desktop platforms, mainly mobile for now. While the graphics won't be the same "quality" when compared, the experience should feel the same, and lights are a great part of the project, being also part of the game design itself.

### Describe the problem or limitation you are having in your project

When in mobile mode inside the editor, or test playing the game on mobile devices, the in game lights flicker, or don't turn on at all. I understand there is such an option in the editor for this limitation, but it's for OpenGL only, which completely undermines whatever work done on the rest of the game graphically so far.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add an option for max lights in the project settings menu.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add the option to change the desired max lights number.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, it's supposed to already be compiled with the engine itself.

### Is there a reason why this should be core and not an add-on in the asset library?

Because this is a core engine option, to be compiled with the engine's code, not changed at runtime.

**Comment by fire** [comment]
Recall that the difference is like 15? for opengl and like thousands for clustered forward rendering. 

So maybe it can be done as a general thing and then we can per platform override. (Per platform configuration override is an existing feature).

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

Unlike Forward+ and Compatibility, the Mobile rendering method was designed in a way that makes it impossible to effectively change how many lights can affect each Mesh resource. This was required for performance reasons, and also to support a broad range of mobile devices (as they have much lower limits compared to desktop platforms).

To avoid issues with lights flickering due to the light count being exceeded:

- Use Distance Fade on light nodes to make distant lights and their shadows fade away. This can also speed up rendering.
- Mix and match omni and spot lights – try to have a 50-50 mix to max out the number of lights you can use. You can make shadow-less spot lights look almost like omni lights by making them have very wide angles (above 90 degrees, up to 179 degrees).
- Split your level mesh into smaller parts – don't use a single mesh for the entire level. This will also improve performance as the level will have more culling opportunities. Don't go overboard though, or the number of draw calls will be excessive (which is very important for mobile devices).
- Use smaller but brighter lights, so that they touch fewer meshes. This also speeds up rendering.
- Use LightmapGI with Light3D nodes that have their light baking mode set to Static instead of Dynamic.

**Comment by luiscesjr** [comment]
> Moving to discussion, as there is no concrete technical implementation provided.
> 
> Unlike Forward+ and Compatibility, the Mobile rendering method was designed in a way that makes it impossible to effectively change how many lights can affect each Mesh resource. This was required for performance reasons, and also to support a broad range of mobile devices (as they have much lower limits compared to desktop platforms).
> 
> To avoid issues with lights flickering due to the light count being exceeded:
> 
> * Use Distance Fade on light nodes to make distant lights and their shadows fade away. This can also speed up rendering.
> * Mix and match omni and spot lights – try to have a 50-50 mix to max out the number of lights you can use. You can make shadow-less spot lights look almost like omni lights by making them have very wide angles (above 90 degrees, up to 179 degrees).
> * Split your level mesh into smaller parts – don't use a single mesh for the entire level. This will also improve performance as the level will have more culling opportunities. Don't go overboard though, or the number of draw calls will be excessive (which is very important for mobile devices).
> * Use smaller but brighter lights, so that they touch fewer meshes. This also speed up rendering.

Thank you for the insight, while I do agree that we have a broad range of devices around, and end user experience should be a priority always, each passing day we get better and better devices. 
I'm using a mid range device for testing, a S20 FE with the infamous exynos inside (The drivers are really bad on this Mali), on parts of my scene that the primitive count goes above the millions it maintains a stable 60 fps (Capped in options). I'd just suggest to at least have that option around, all in all would be a nice feature to be able to tweak around. As all options are, optional =)
Since there is already the possibility to change that in the editor, but on the opengl side. 

And also thanks for the tips, I'll work on dividing my levels into smaller chunks, since that'll give for sure the most immediate results, and also the light fading option.
